oauth2/src/main/java/com/linecorp/armeria/common/auth/oauth2/OAuth2TokenDescriptor.java

```diff
@@ -0,0 +1,511 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.auth.oauth2;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.Serializable;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * Defines a structure of the Token Introspection Response, as per
+ * <a href="https://tools.ietf.org/html/rfc7662#section-2.2">[RFC7662], Section 2.2</a>.
+ */
+public class OAuth2TokenDescriptor implements Serializable {
+
+    private static final long serialVersionUID = -3976877781134216467L;
+
+    /**
+     * Creates a new {@link OAuth2TokenDescriptor} based on the {@code JSON}-formatted raw response body.
+     * @param rawResponse {@code JSON}-formatted raw response body.
+     * @return A new instance of {@link OAuth2TokenDescriptor}.
+     */
+    public static OAuth2TokenDescriptor of(String rawResponse) {
+        return OAuth2TokenDescriptorBuilder.of(rawResponse);
+    }
+
+    /**
+     * Creates a new {@link OAuth2TokenDescriptorBuilder} to build a new {@link OAuth2TokenDescriptor} and
+     * supplied it with a value of {@code active} Token Introspection Response field.
+     * @return A new instance of {@link OAuth2TokenDescriptorBuilder}.
+     */
+    public static OAuth2TokenDescriptorBuilder builder(boolean active) {
+        return new OAuth2TokenDescriptorBuilder(active);
+    }
+
+    static final String ACTIVE = "active";
+    static final String SCOPE = "scope";
+    static final String CLIENT_ID = "client_id";
+    static final String USER_NAME = "username";
+    static final String TOKEN_TYPE = "token_type";
+    static final String EXPIRES_AT = "exp";
+    static final String ISSUED_AT = "iat";
+    static final String NOT_BEFORE = "nbf";
+    static final String SUBJECT = "sub";
+    static final String AUDIENCE = "aud";
+    static final String ISSUER = "iss";
+    static final String JWT_ID = "jti";
+
+    static final String SCOPE_SEPARATOR = " ";
+
+    static final ObjectMapper JSON = new ObjectMapper();
```

Comment:
Could share one `ObjectMapper` instance across all classes unless some needs to be configured differently?

Reply:
fixed